#reads filtering and genome assemble

## 1.reads filter
###1.1 quality filter
    filter reads with “N”s that constitute more than 5 percent of nucleotides  
    filter reads that have less than or equal to 65 percent of nucleotides with quality scores less than or equal to 7  
###1.2 lighter  
    trimming the Short-insert paired-end Illumina libraries (<2kbp)
    #lighter v1.0.7
    ```
    lighter -r scriptfilter.500bp.1.filter.fq.gz -r scriptfilter.500bp.2.filter.fq.gz -k 17 600000000 0.156 -t 20 -trim
    ```
###1.3 fastuniq
    removal of duplicates in long-insert paired-end Illumina libraries (>=2kb)
    #FastUniq v1.1
    ```
    fastuniq -i ip_2kbp.txt -t q -o op_2kbp.1.fq -p op_2kbp.2.fq
    ```
## 2.assemble
##2.1 platanus
    First, Reads with 500bp and 800bp insert size were used to constructs contigs using the algorithm based on de Bruijn graph.  
    Second, the order of contigs is determined according to paired-end (mate-pair) data and all reads were used to constructs scaffolds.   
    Third, paired-end reads localized on gaps in scaffolds were assembled and gaps were closed.
    #platanus v1.2.1
    #contig
    ```
    platanus assemble  -o 01.contig -f 01.CleanReads/scriptfilter.500bp.1.filter.cor.fq 01.CleanReads/scriptfilter.500bp.2.filter.cor.fq 01.CleanReads/scriptfilter.800bp.1.filter.cor.fq 01.CleanReads/scriptfilter.800bp.2.filter.cor.fq -t 50 -m 500
    #result: 01.contig_contig.fa
    ```
    #scaffold:
    ```
    platanus scaffold -o 02.scaffold -b 01.contig_contigBubble.fa -c 01.contig_contig.fa -t 50 -IP1 01.CleanReads/scriptfilter.500bp.1.filter.cor.fq 01.CleanReads/scriptfilter.500bp.2.filter.cor.fq -IP2 01.CleanReads/scriptfilter.800bp.1.filter.cor.fq 01.CleanReads/scriptfilter.800bp.2.filter.cor.fq -OP3 01.CleanReads/op_2kbp.1.fq 01.CleanReads/op_2kbp.2.fq -OP4 01.CleanReads/op_5kbp.1.fq 01.CleanReads/op_5kbp.2.fq -OP5 01.CleanReads/op_10kbp.1.fq 01.CleanReads/op_10kbp.2.fq
    #result: 02.scaffold_scaffold.fa
    ```
    #gap_close:
    platanus gap_close -c 02.scaffold_scaffold.fa -o 03.platanusGapClose -t 50 -IP1 01.CleanReads/scriptfilter.500bp.1.filter.cor.fq 01.CleanReads/scriptfilter.500bp.2.filter.cor.fq -IP2 01.CleanReads/scriptfilter.800bp.1.filter.cor.fq 01.CleanReads/scriptfilter.800bp.2.filter.cor.fq -OP3 01.CleanReads/op_2kbp.1.fq 01.CleanReads/op_2kbp.2.fq -OP4 01.CleanReads/op_5kbp.1.fq 01.CleanReads/op_5kbp.2.fq -OP5 01.CleanReads/op_10kbp.1.fq 01.CleanReads/op_10kbp.2.fq
    #result：03.platanusGapClose_gapClosed.fa
    ```
##2.2 Soap  
    gap_close
    ##SOAPec v2.02
    ```
    ~/software/GapCloser/GapCloser -a 03.platanusGapClose_gapClosed.fa -b 04.soapGapClose.config -o 04.soapGapClose.fa -t 50 -l 100
    #result: 04.soapGapClose.fa
##2.3 filter the reads whitch length less than 200bp (N was exclude)
    ```
#! /usr/bin/perl -w
#此脚本用于基因组gap close后再过滤，去掉不包括N的200bp及以下的序列,修改第8行$filter_length可以改变过滤碱基数的大小
#脚本路径 输入文件 输出文件
use strict;
use Bio::SeqIO;
my $file=shift;
my $op=shift;
my $filter_length = 200;
open (O,">$op")||die("$!\n");
my $fa = Bio::SeqIO->new(-file=>$file, -format=>'fasta');
while(my $seq= $fa->next_seq){
    my $id=$seq->id;
    my $seq=$seq->seq;
    my $N=($seq=~tr/[N|n]//);
    my $scaffold_length=length $seq;
    my $effective_scaffold_length=$scaffold_length-$N;
    if($effective_scaffold_length>$filter_length){
      print O ">$id\n$seq\n";
    }
}
close O;
```
